### ISSUE TITLE
restore() && reset() non-functional if wrapping all object methods

### ISSUE THREAD

**Issue by olirogers** [sinonjs/sinon#2192]
**Describe the bug**
When using `sinon.spy(<obj>)` the method `sinon.restore()` does not restore the object back to the original state.

This is very useful for mocha (other) tests which use beforeEach/afterEach to restore the object to previous states.

A workaround I'm using is `sinon.spy(<obj>, "method_name")` as this works as expected and `sinon.restore()` removes the wrapped methods.

**To Reproduce**
See [https://runkit.com/olirogers/sinon-restore-issue/2.0.0](https://runkit.com/olirogers/sinon-restore-issue/2.0.0)

1. Wrap an object
2. (Optionally) run some tests on it
3. Restore the sandbox/sinon
4. Wrap the object again
5. Receive an error that `Attempted to wrap <method_name> which is already wrapped`

**Expected behavior**
Script to run successfully, restore removes the wrapping of the methods on the object.

**Screenshots**
N/A - See [https://runkit.com/olirogers/sinon-restore-issue/2.0.0](https://runkit.com/olirogers/sinon-restore-issue/2.0.0)

**Context (please complete the following information):**

* Library version: 8.0.2
* Environment: Node 10.18.0 (from runkit)
* Example URL: [https://runkit.com/olirogers/sinon-restore-issue/2.0.0](https://runkit.com/olirogers/sinon-restore-issue/2.0.0)
* Other libraries you are using: @sinonjs/referee (version 4.0.0)

**Additional context**
This issue was encountered following the official guides at - https://sinonjs.org/releases/latest/spies/, version 1.0.0. of the runkit has an example that should also exhibit this bug (but runkit was being slow!).

The same behaviour is experienced with `sinon.reset()`. It appears to be calling reset on the object itself (`new Foo{}`) rather than each of the methods.

I found #1553 but assumed that this should be called automatically from the sandbox/environment restore()?

Let me know if you need any more information.

Oli

**Comment by fatso83** [comment]
Excellent bug report! This seems weird if this is not working as expected, so I would first try to see if this  is covered by some existing test and why it does not work.

**Comment by fatso83** [comment]
By playing in the console I have verified that restoring _stubs_ work, but _spies_ are not restored.

**Comment by fatso83** [comment]
Spies and stubs mutate objects in mostly the same ways, so I don't see why we treat the two so differently. Stubs have a much higher branching factor:
```javascript
    sandbox.spy = function spy() {
        var s = sinonSpy.apply(sinonSpy, arguments);

        push(collection, s);

        return s;
    };

    sandbox.stub = function stub(object, property) {
        if (isEsModule(object)) {
            throw new TypeError("ES Modules cannot be stubbed");
        }

        if (isNonExistentOwnProperty(object, property)) {
            throw new TypeError("Cannot stub non-existent own property " + valueToString(property));
        }

        var stubbed = sinonStub.apply(null, arguments);
        var isStubbingEntireObject = typeof property === "undefined" && typeof object === "object";

        if (isStubbingEntireObject) {
            var ownMethods = collectOwnMethods(stubbed);

            forEach(ownMethods, function(method) {
                push(collection, method);
            });

            usePromiseLibrary(promiseLib, ownMethods);
        } else {
            push(collection, stubbed);
            usePromiseLibrary(promiseLib, stubbed);
        }

        return stubbed;
    };
```

AFAIK, most of this code could/should be shared as it applies to both.

**Comment by mroderick** [comment]
The fix has been published in `sinon@8.0.4`

**Comment by olirogers** [comment]
Thank you for looking at this quickly! It helps greatly :)
